pkg/metrics/metrics.go

```diff
@@ -0,0 +1,276 @@
+// CounterVec has already been registered, the existing metric will be returned.
+func registerCounterVec(r prometheus.Registerer, g *prometheus.CounterVec) (*prometheus.CounterVec, error) {
+	c, err := registerCollector(r, g)
+	if err != nil {
+		return nil, err
+	}
+	if gv, ok := c.(*prometheus.CounterVec); ok {
+		return gv, nil
+	}
+	return nil, errors.New("failed to register counter vec")
+}
+
+// registerHistogramVec registers the Histogram with the provided Registerer. If the
+// HistogramVec has already been registered, the existing metric will be returned.
+func registerHistogramVec(r prometheus.Registerer, h *prometheus.HistogramVec) (*prometheus.HistogramVec, error) {
+	c, err := registerCollector(r, h)
+	if err != nil {
+		return nil, err
+	}
+	if v, ok := c.(*prometheus.HistogramVec); ok {
+		return v, nil
+	}
+	return nil, errors.New("failed to register historgram vec")
+}
+
+// registerGaugeVec registers the GaugeVec with the provided Registerer. If the
+// GaugeVec has already been registered, the existing GaugeVec will be returned.
+func registerGaugeVec(r prometheus.Registerer, g *prometheus.GaugeVec) (*prometheus.GaugeVec, error) {
+	c, err := registerCollector(r, g)
+	if err != nil {
+		return nil, err
+	}
+	if gv, ok := c.(*prometheus.GaugeVec); ok {
+		return gv, nil
+	}
+	return nil, errors.New("failed to register gauge vec")
+}
+
+// registerCounter registers the Counter with the provided Registerer. If the
+// counter has already been registered, the existing metric will be returned.
+func registerCounter(r prometheus.Registerer, cntr prometheus.Counter) (prometheus.Counter, error) {
+	c, err := registerCollector(r, cntr)
+	if err != nil {
+		return nil, err
+	}
+	if rc, ok := c.(prometheus.Counter); ok {
+		return rc, nil
+	}
+	return nil, errors.New("failed to register counter")
+}
+
+// registerHistogram registers the Histogram with the provided Registerer. If the
+// histogram has already been registered, the existing metric will be returned.
+func registerHistogram(r prometheus.Registerer, h prometheus.Histogram) (prometheus.Histogram, error) {
+	c, err := registerCollector(r, h)
+	if err != nil {
+		return nil, err
+	}
+
+	if rh, ok := c.(prometheus.Histogram); ok {
+		return rh, nil
+	}
+
+	return nil, errors.New("failed to register histogram")
+}
+
+// registerGauge registers the Gauge with the provided Registerer. If the
+// gauge has already been registered, the existing metric will be returned.
+func registerGauge(r prometheus.Registerer, g prometheus.Gauge) (prometheus.Gauge, error) {
+	c, err := registerCollector(r, g)
+	if err != nil {
+		return nil, err
+	}
+	if rg, ok := c.(prometheus.Gauge); ok {
+		return rg, nil
+	}
+	return nil, errors.New("failed to register gauge")
+}
+
+// registerCollector is an helper to register a metric and log registration errors
+func registerCollector(r prometheus.Registerer, c prometheus.Collector) (prometheus.Collector, error) {
```

Comment:
I suggest renaming this helper to be `registerMetric`.

You might consider naming it `registerMetricOrDie` (idiom used elsewhere), dropping the error return, and instead of returning an error, panic here. That simplifies the callers. I would likewise push the panic down in the the per-type helpers when they fail the type check.

Reply:
Sure, @ewhamilton. This makes sense. I've simplified the APIs by pushing the panic down to the per-type helpers if they fail type check and made changes to the now `reigsterMetricOrDie` method.